src/test/orphanage_tests.cpp

```diff
@@ -162,22 +162,53 @@ BOOST_AUTO_TEST_CASE(DoS_mapOrphans)
         BOOST_CHECK(!orphanage.AddTx(MakeTransactionRef(tx), i));
     }
 
-    // Test EraseOrphansFor:
+    size_t expected_num_orphans = orphanage.CountOrphans();
+
+    // Non-existent peer; nothing should be deleted
+    orphanage.EraseForPeer(/*peer=*/-1);
+    BOOST_CHECK_EQUAL(orphanage.CountOrphans(), expected_num_orphans);
+
+    // Test EraseOrphansFor: Each of first three peers announced
+    // two transactions each.
     for (NodeId i = 0; i < 3; i++)
     {
-        size_t sizeBefore = orphanage.CountOrphans();
         orphanage.EraseForPeer(i);
-        BOOST_CHECK(orphanage.CountOrphans() < sizeBefore);
+        BOOST_CHECK(orphanage.CountOrphans() == expected_num_orphans - 2);
+        expected_num_orphans -= 2;
     }
 
-    // Test LimitOrphanTxSize() function:
+    // Test LimitOrphanTxSize() function, nothing should timeout:
     FastRandomContext rng{/*fDeterministic=*/true};
+    orphanage.LimitOrphans(/*max_orphans=*/expected_num_orphans, rng);
+    BOOST_CHECK_EQUAL(orphanage.CountOrphans(), expected_num_orphans);
+    orphanage.LimitOrphans(/*max_orphans=*/expected_num_orphans - 1, rng);
+    BOOST_CHECK_EQUAL(orphanage.CountOrphans(), expected_num_orphans - 1);
     orphanage.LimitOrphans(40, rng);
-    BOOST_CHECK(orphanage.CountOrphans() <= 40);
+    BOOST_CHECK_EQUAL(orphanage.CountOrphans(), 40);
     orphanage.LimitOrphans(10, rng);
-    BOOST_CHECK(orphanage.CountOrphans() <= 10);
+    BOOST_CHECK_EQUAL(orphanage.CountOrphans(), 10);
     orphanage.LimitOrphans(0, rng);
-    BOOST_CHECK(orphanage.CountOrphans() == 0);
+    BOOST_CHECK_EQUAL(orphanage.CountOrphans(), 0);
+
+    // Add one more orphan, check timeout logic
+    CMutableTransaction tx;
+    tx.vin.resize(1);
+    tx.vin[0].prevout.n = 0;
+    tx.vin[0].prevout.hash = Txid::FromUint256(InsecureRand256());
+    tx.vin[0].scriptSig << OP_1;
+    tx.vout.resize(1);
+    tx.vout[0].nValue = 1*CENT;
+    tx.vout[0].scriptPubKey = GetScriptForDestination(PKHash(key.GetPubKey()));
+
+    orphanage.AddTx(MakeTransactionRef(tx), 0);
+    orphanage.LimitOrphans(1, rng);
+    BOOST_CHECK_EQUAL(orphanage.CountOrphans(), 1);
+
+    // Jump ORPHAN_TX_EXPIRE_TIME minutes, orphan should be timed out on limiting
+    SetMockTime(GetTime() + 20 * 60);
+    BOOST_CHECK_EQUAL(orphanage.CountOrphans(), 1);
+    orphanage.LimitOrphans(1, rng);
+    BOOST_CHECK_EQUAL(orphanage.CountOrphans(), 0);
```

Comment:
Maybe out of scope but I wonder if it would be best to add a `current_time` parameter to the `TxOrphanage` methods to make it easier to test and fuzz the expiration (also should have a test for the sorting for `GetChildrenFromSamePeer` results)

Reply:
probably a good idea, though you might have better opinions on orphanage interface. I tightened the time testing a bit more by setting mocktime at beginning, then checking boundary conditions via subsequent setting of the time.